File: platforms/ios/foundation/Sources/NativeblocksFoundation/modifier/Width.swift
import NativeblocksCompiler
import NativeblocksRuntime
import SwiftUI

@Modifier(
    name: "Width",
    keyType: "nativeblocks/width",
    description: "Sets the width of the block.",
    version: 1,
    versionName: "1"
)
struct Width: ViewModifier {

    @ModifierData(description: "The width of the block ('infinity', 'fit' or a number).", defaultValue: "fit")
    var value: String = "fit"

    @ViewBuilder
    func body(content: Content) -> some View {
        switch value {
        case "infinity":
            content.frame(maxWidth: .infinity)
        case "fit":
            content
        default:
            if let width = Double(value) {
                content.frame(width: width)
            } else {
                content
            }
        }
    }
}

File: platforms/android/foundation/foundation/src/main/java/io/nativeblocks/foundation/modifier/Width.kt
package io.nativeblocks.foundation.modifier

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.wrapContentWidth
import androidx.compose.ui.unit.dp
import io.nativeblocks.compiler.type.Modifier
import io.nativeblocks.compiler.type.ModifierData
import androidx.compose.ui.Modifier as UiModifier

@Modifier(
    keyType = "nativeblocks/width",
    name = "Width",
    description = "Sets the width of the block.",
    version = 1,
    versionName = "1",
)
internal fun width(
    @ModifierData(
        description = "The width of the block ('match', 'wrap' or a number in DP).",
        defaultValue = "wrap",
    ) value: String = "wrap",
): UiModifier = when (value) {
    "match" -> UiModifier.fillMaxWidth()
    "wrap" -> UiModifier.wrapContentWidth()
    else -> {
        val dpValue = value.toFloatOrNull()
        if (dpValue != null) UiModifier.width(dpValue.dp) else UiModifier
    }
}
